include/nbl/asset/IPreHashed.h

```diff
@@ -41,36 +41,31 @@ class IPreHashed : public IAsset
 
 		static inline void discardDependantsContents(const std::span<IAsset*> roots)
 		{
-			struct stack_entry_t
-			{
-				const IAsset* asset;
-				core::unordered_set<const IAsset*> unvisitedChilds;
-			};
-			core::stack<stack_entry_t> stack;
-			core::unordered_set<const IAsset*> alreadyVisited;
-			auto push = [&stack,&alreadyVisited](const IAsset* node) -> void
+			core::stack<IAsset*> stack;
+			core::unordered_set<IAsset*> alreadyVisited; // whether we have push the node to the stack
+			core::unordered_set<IAsset*> alreadyDescended; // whether we have push the children to the stack
+			auto push = [&stack,&alreadyVisited](IAsset* node) -> void
 			{
 				if (!node)
 					return;
 				const auto [dummy,inserted] = alreadyVisited.insert(node);
 				if (inserted)
-					stack.push({ .asset = node, .unvisitedChilds = node->computeDependants()});
+					stack.push(node);
 			};
 			for (const auto& root : roots)
 				push(root);
 			while (!stack.empty())
 			{
-				auto& entry = stack.top();
-				if (entry.unvisitedChilds.size() > 0)
+				auto* entry = stack.top();
+				const auto [dummy, inserted] = alreadyDescended.insert(entry);
+				if (inserted)
 				{
-					auto dep = *entry.unvisitedChilds.begin();
-					entry.unvisitedChilds.erase(entry.unvisitedChilds.begin());
-					push(dep);
-				}
-				else
+          core::unordered_set<IAsset*> dependants = entry->computeDependants();
+					for (auto* dependant : dependants) push(dependant);
```

Comment:
wondering about one thing, maybe our `push` should just be lambda that new `visitDpendants` should take